Changes to build and runit on Windows 10 with WSL (Ubuntu)

TenEniunlsl opened the pull request:
Issues with GtkSharp packages
WebkitGtkSharp not found on nuget.org or something
Android build deactivated
Project references added

TenEniunlsl committed:
Changes to run it on Windows 10 with WSL (Ubuntu)

Lemon73-Computing commented on Directory.Build.Override.props:
```diff
@@ -7,7 +7,7 @@
     <_IncludeTizen></_IncludeTizen>
     <_IncludeGtk>true</_IncludeGtk>
-    <_IncludeAndroid>true</_IncludeAndroid>
+    <_IncludeAndroid>false</_IncludeAndroid>
```
Making this change may cause CI/CD to cease functioning.

Therefore, I believe it would be preferable either to modify the CI/CD setup accordingly

https://github.com/MauiGtk/maui-linux/blob/118de7cfdb2dc4b918b261d379ab4a089c78d958/.github/workflows/build-gtk.yml#L38

```diff
# .github/workflows/build-gtk.yml
- sed -i 's/_IncludeAndroid>true/_IncludeAndroid>/g' Directory.Build.Override.props 
+ sed -i 's/_IncludeAndroid>/_IncludeAndroid>/g' Directory.Build.Override.props
```

or revert this change.

TenEniunlsl replied:
Ok i changed it because VS 2022 would'nt like to install an old Android SDK or something

Lemon73-Computing replied:
Okay, could you update the CI/CD configuration in `.github/workflows/build-gtk.yml` accordingly? The required changes are as previously discussed.

Lemon73-Computing commented on NuGet.config:
```diff
@@ -6,6 +6,7 @@
     <add key="dotnet-eng" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-eng/nuget/v3/index.json" protocolVersion="3" />
     <add key="dotnet-public" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-public/nuget/v3/index.json" protocolVersion="3" />
     <add key="benchmark-dotnet-prerelease" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/benchmark-dotnet-prerelease/nuget/v3/index.json" />
+    <add key="Local" value="C:\source\GtkSharp\BuildOutput\NugetPackages" />
```
As it uses an absolute path specification, it may not work on other people's computers?

TenEniunlsl replied:
The problem here was that WebkitGtkSharp isn't on nuget.org or similar sources so i had to compile and provide it my self

TenEniunlsl replied:
Maybe i had later seen Nuget.org was deactivated in Visual Studio but i'm not sure about it

TenEniunlsl replied:
ok i have found it, nuget.org isn't in NuGet.config, i will change it

Lemon73-Computing commented on src/BlazorWebView/src/Maui/Microsoft.AspNetCore.Components.WebView.Maui.csproj:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <TargetFrameworks>$(_MauiDotNetTfm);$(MauiPlatforms)</TargetFrameworks>
+    <TargetFrameworks>$(_MauiDotNetTfm)</TargetFrameworks>
```
What is the reason for deleting the platform?

TenEniunlsl replied:
Something like a mismatch from compiler, i can try to reproduce it and show you the error message

TenEniunlsl replied:
In Visual Studio i got messages like that
<img width="2050" height="277" alt="image" src="https://github.com/user-attachments/assets/d691408d-afea-4e10-bccc-5daf720a9b6f" />

thats the Text, sry it's in German and everyone it's the same with another Project, you can scroll down there is one i have translated


```
Problem beim Projekt "Graphics.Skia" während des letzten Vorgangs: Der Wert der Eigenschaft "TargetFrameworks" ist zwischen Zielframeworks inkonsistent. Diese Eigenschaft muss identisch sein, damit die NuGet-Wiederherstellung ordnungsgemäß funktioniert. Der Wert "netstandard2.1;netstandard2.0;net8.0;net8.0-gtk;;net7.0;" aus der Debug|AnyCPU|netstandard2.1-Konfiguration wird verwendet, andere Zielframeworks wählen möglicherweise keine NuGet-Objekte aus.. Sie müssen möglicherweise die Projektmappe neu laden, nachdem Sie das Problem behoben haben. 
Problem beim Projekt "Maps" während des letzten Vorgangs: Der Wert der Eigenschaft "TargetFrameworks" ist zwischen Zielframeworks inkonsistent. Diese Eigenschaft muss identisch sein, damit die NuGet-Wiederherstellung ordnungsgemäß funktioniert. Der Wert "netstandard2.1;netstandard2.0;net8.0;net8.0-gtk;" aus der Debug|AnyCPU|netstandard2.1-Konfiguration wird verwendet, andere Zielframeworks wählen möglicherweise keine NuGet-Objekte aus.. Sie müssen möglicherweise die Projektmappe neu laden, nachdem Sie das Problem behoben haben. 
Problem beim Projekt "Controls.Xaml" während des letzten Vorgangs: Der Wert der Eigenschaft "TargetFrameworks" ist zwischen Zielframeworks inkonsistent. Diese Eigenschaft muss identisch sein, damit die NuGet-Wiederherstellung ordnungsgemäß funktioniert. Der Wert "netstandard2.1;netstandard2.0;net8.0;net8.0-gtk;net8.0-android;" aus der Debug|AnyCPU|netstandard2.1-Konfiguration wird verwendet, andere Zielframeworks wählen möglicherweise keine NuGet-Objekte aus.. Sie müssen möglicherweise die Projektmappe neu laden, nachdem Sie das Problem behoben haben. 
Problem beim Projekt "Controls.Foldable" während des letzten Vorgangs: Der Wert der Eigenschaft "TargetFrameworks" ist zwischen Zielframeworks inkonsistent. Diese Eigenschaft muss identisch sein, damit die NuGet-Wiederherstellung ordnungsgemäß funktioniert. Der Wert "net8.0;net8.0-gtk;net8.0-android;" aus der Debug|AnyCPU|net8.0-Konfiguration wird verwendet, andere Zielframeworks wählen möglicherweise keine NuGet-Objekte aus.. Sie müssen möglicherweise die Projektmappe neu laden, nachdem Sie das Problem behoben haben. 
Problem beim Projekt "Microsoft.AspNetCore.Components.WebView.Maui" während des letzten Vorgangs: Der Wert der Eigenschaft "TargetFrameworks" ist zwischen Zielframeworks inkonsistent. Diese Eigenschaft muss identisch sein, damit die NuGet-Wiederherstellung ordnungsgemäß funktioniert. Der Wert "net8.0;net8.0-gtk;" aus der Debug|AnyCPU|net8.0-Konfiguration wird verwendet, andere Zielframeworks wählen möglicherweise keine NuGet-Objekte aus.. Sie müssen möglicherweise die Projektmappe neu laden, nachdem Sie das Problem behoben haben. 
Problem beim Projekt "Maui.Controls.Sample" während des letzten Vorgangs: Der Wert der Eigenschaft "TargetFrameworks" ist zwischen Zielframeworks inkonsistent. Diese Eigenschaft muss identisch sein, damit die NuGet-Wiederherstellung ordnungsgemäß funktioniert. Der Wert "net8.0-gtk;net8.0-android;;net8.0" aus der Debug|AnyCPU|net8.0-gtk-Konfiguration wird verwendet, andere Zielframeworks wählen möglicherweise keine NuGet-Objekte aus.. Sie müssen möglicherweise die Projektmappe neu laden, nachdem Sie das Problem behoben haben. 
Problem beim Projekt "Compatibility" während des letzten Vorgangs: Der Wert der Eigenschaft "TargetFrameworks" ist zwischen Zielframeworks inkonsistent. Diese Eigenschaft muss identisch sein, damit die NuGet-Wiederherstellung ordnungsgemäß funktioniert. Der Wert "netstandard2.1;netstandard2.0;net8.0;net8.0-gtk;net8.0-android;" aus der Debug|AnyCPU|netstandard2.1-Konfiguration wird verwendet, andere Zielframeworks wählen möglicherweise keine NuGet-Objekte aus.. Sie müssen möglicherweise die Projektmappe neu laden, nachdem Sie das Problem behoben haben. 
Problem beim Projekt "Core" während des letzten Vorgangs: Der Wert der Eigenschaft "TargetFrameworks" ist zwischen Zielframeworks inkonsistent. Diese Eigenschaft muss identisch sein, damit die NuGet-Wiederherstellung ordnungsgemäß funktioniert. Der Wert "netstandard2.1;netstandard2.0;net8.0;net8.0-gtk;net8.0-android;" aus der Debug|AnyCPU|netstandard2.1-Konfiguration wird verwendet, andere Zielframeworks wählen möglicherweise keine NuGet-Objekte aus.. Sie müssen möglicherweise die Projektmappe neu laden, nachdem Sie das Problem behoben haben. 
Problem beim Projekt "Graphics" während des letzten Vorgangs: Der Wert der Eigenschaft "TargetFrameworks" ist zwischen Zielframeworks inkonsistent. Diese Eigenschaft muss identisch sein, damit die NuGet-Wiederherstellung ordnungsgemäß funktioniert. Der Wert "netstandard2.1;netstandard2.0;net8.0;net8.0-gtk;net8.0-android;;net7.0;net7.0-android;" aus der Debug|AnyCPU|netstandard2.1-Konfiguration wird verwendet, andere Zielframeworks wählen möglicherweise keine NuGet-Objekte aus.. Sie müssen möglicherweise die Projektmappe neu laden, nachdem Sie das Problem behoben haben. 
Problem beim Projekt "Essentials" während des letzten Vorgangs: Der Wert der Eigenschaft "TargetFrameworks" ist zwischen Zielframeworks inkonsistent. Diese Eigenschaft muss identisch sein, damit die NuGet-Wiederherstellung ordnungsgemäß funktioniert. Der Wert "netstandard2.1;netstandard2.0;net8.0;net8.0-gtk;net8.0-android;;net7.0;net7.0-android;" aus der Debug|AnyCPU|netstandard2.1-Konfiguration wird verwendet, andere Zielframeworks wählen möglicherweise keine NuGet-Objekte aus.. Sie müssen möglicherweise die Projektmappe neu laden, nachdem Sie das Problem behoben haben. 
Problem beim Projekt "Controls.Core" während des letzten Vorgangs: Der Wert der Eigenschaft "TargetFrameworks" ist zwischen Zielframeworks inkonsistent. Diese Eigenschaft muss identisch sein, damit die NuGet-Wiederherstellung ordnungsgemäß funktioniert. Der Wert "netstandard2.1;netstandard2.0;net8.0;net8.0-gtk;net8.0-android;" aus der Debug|AnyCPU|netstandard2.1-Konfiguration wird verwendet, andere Zielframeworks wählen möglicherweise keine NuGet-Objekte aus.. Sie müssen möglicherweise die Projektmappe neu laden, nachdem Sie das Problem behoben haben.

```

Issue in Graphics.Skia project during last operation: The value of the TargetFrameworks property is inconsistent between target frameworks. This property must be the same for NuGet restore to work properly. The value "netstandard2.1;netstandard2.0;net8.0;net8.0-gtk;;net7.0;" from the Debug|AnyCPU|netstandard2.1 configuration is used, other target frameworks may not select NuGet objects. You may need to reload the solution after resolving the issue.

TenEniunlsl replied:
Any suggestions?

Lemon73-Computing replied:
I'm sorry for the delayed response.

I was developing within a Docker environment, and it is functioning without issue even with the 'Platform'. The Dockerfile I was using may be of some reference: <https://github.com/MauiGtk/maui-docker>

TenEniunlsl replied:
Ok but what should we do here? I have now 5 PRs (practically, but not testet yet) and all waiting for that

Lemon73-Computing replied:
I think `;$(MauiPlatforms)` should be kept, since it’s also used on platforms other than `gtk` (e.g., `windows`, `macos`, etc.).

In the `maui-docker` project we’re developing, it works with `;$(MauiPlatforms)`, so it should be possible.

<!--
If you can resolve it alone, I can merge this branch. And you would be better to `rebase` the other five branches.

If it seems difficult, I’d like to give it a try myself as well. However, I’m currently busy, so it may take some time before I can review it in detail.
-->

Lemon73-Computing commented on src/Graphics/src/Graphics.Gtk/Graphics.Gtk.csproj:
```diff
@@ -9,11 +9,16 @@
   <ItemGroup>
     <ProjectReference Include="..\Graphics\Graphics.csproj" />
   </ItemGroup>
-  <ItemGroup Condition="$(TargetFramework.Contains('-gtk')) != 'true'">
```
What is the reason for removing the platform specification here as well?

TenEniunlsl replied:
i have asked the same question [here](https://github.com/Lemon73-Computing/maui-linux/issues/28#issuecomment-3375006045) 

`<ItemGroup Condition="$(TargetFramework.Contains('-gtk')) != 'true'">
<PackageReference Include="GtkSharp" />
</ItemGroup>`

For me it doesn't make sense to include GtkSharp if it isn't a -gtk project and also it would'nt compile

TenEniunlsl replied:
In my opinion it should look like

`<ItemGroup Condition="$(TargetFrameworks.Contains('-gtk')) == 'True'">
  <PackageReference Include="CairoSharp" Version="3.24.24.95" />
  <PackageReference Include="GtkSharp" Version="3.24.24.95" />
  <PackageReference Include="GdkSharp" Version="3.24.24.95" />
  <PackageReference Include="PangoSharp" Version="3.24.24.95" />
</ItemGroup>`

There was n s missing in TargetFrameworks and it should be "== True", correct me if i'm wrong

Lemon73-Computing replied:
I looked into it on my end, and it seems that [MSBuild is case-insensitive](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-conditions?view=vs-2022#:~:text=for%20empty%20values.-,This%20check%20is%20case%20insensitive.,-%27stringA%27%20!%3D%20%27stringB%27).

TenEniunlsl replied:
maybe i will try it, as i changed the workload i had read it is case sensitive 

but it should'nt != it should ==

Commit:
Changes to Load Files, embeded or written to output directory with the changed GTK Workload

Lemon73-Computing reviewed (commented):
Thank you for contribution!

Just to confirm, is this PR ready for review, or are you still making changes?

Commit:
Misspelling and Condition fixed
Condition="$(TargetFrameworks.Contains('-gtk')) == 'True'"

Nuget.org added to NuGet.config

TenEniunlsl committed:
Merge branch 'MauiGtk:main' into main

Lemon73-Computing commented on src/Controls/samples/Controls.Sample/Maui.Controls.Sample.csproj:
```diff
@@ -39,24 +39,24 @@
-    <PackageReference Include="Microsoft.Extensions.DependencyInjection" />
+    <PackageReference Include="Microsoft.Extensions.Configuration" Version="8.0.0" />
+    <PackageReference Include="Microsoft.Extensions.Logging" Version="8.0.0" />
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="8.0.0" />
```
Package versions are already defined in `NuGetVersions.targets` and `Versions.props`, so I believe an explicit version specification here is unnecessary.

https://github.com/MauiGtk/maui-linux/blob/118de7cfdb2dc4b918b261d379ab4a089c78d958/eng/NuGetVersions.targets#L99-L102

https://github.com/MauiGtk/maui-linux/blob/118de7cfdb2dc4b918b261d379ab4a089c78d958/eng/Versions.props#L17

Lemon73-Computing commented on src/Essentials/src/Essentials.csproj:
```diff
@@ -70,6 +70,10 @@
     <Compile Include="**\*.Gtk.*.cs" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder)" />
   </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="GdkSharp" Version="3.24.24.95" />
```
Are these packages necessary? I think the `GTK workload` should be sufficient for this to work. (also: `src/Core/src/Core.csproj`)
